### PR TITLE
Add a host location property to TableProperties

### PR DESCRIPTION
**Pull request by anand1976**
This PR adds support for writing a location identifier of the DB host to SST files as a table property. By default, the hostname is used, but can be overridden by the user. There have been some recent corruptions in files written by ```SstFileWriter``` before checksumming, so this property can be used to trace it back to the writing host and checking the host for hardware isues.

Test plan:
Add new unit tests

### CODE REVIEW

File: include/rocksdb/options.h [hunk 1]
```diff
@@ -1157,6 +1157,13 @@ struct DBOptions {
   //
   // Default: 1000000 (microseconds).
   uint64_t bgerror_resume_retry_interval = 1000000;
+
+  // A string identifying the location of the machine hosting the DB. This
+  // will be written as a property in every SST file written by the DB (or
+  // by offline writers such as SstFileWriter and RepairDB)
+  //
+  // Default: hostname
+  std::string db_host_location;
```
Comment (mrambacher) [inline include/rocksdb/options.h:1166]: Can you give a better comment as to what this would be used for?  Why would a user want to set this and why do we want this in the SST files?

**Review by mrambacher — COMMENTED** [review]
I am a bit worried about how this change will behave for options.  I know we attempt to validate that the options have not changed from what was set in the file but it seems like that could very well happen here (if the option was not set on startup).  It would also make an Options file more tied to a specific host (if we care).

There also needs to be changes to options_settable_test.

**Comment by anand1976** [comment]
@mrambacher 

> I am a bit worried about how this change will behave for options. I know we attempt to validate that the options have not changed from what was set in the file but it seems like that could very well happen here (if the option was not set on startup). It would also make an Options file more tied to a specific host (if we care).

I didn't understand the concern. We do sanitize other options, such as ```wal_dir```, before writing out to the OPTIONS file.  I don't think we read back the options file and try to validate.

File: HISTORY.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
...
 * AdvancedColumnFamilyOptions::force_consistency_checks is now true by default, for more proactive DB corruption detection at virtually no cost (estimated two extra CPU cycles per million on a major production workload). Corruptions reported by these checks now mention "force_consistency_checks" in case a false positive corruption report is suspected and the option needs to be disabled (unlikely). Since existing column families have a saved setting for force_consistency_checks, only new column families will pick up the new default.
+* A new option, db_host_location, has been added to DBOptions. If left empty, this option defaults to the hostname (FQDN). Either the default or the user specified string is written to SST files as a table property
```
[the remark this answers is not in the document]
Comment (anand1976) [reply HISTORY.md:11]: Ok, we can do that. I'm not keen on introducing another option field to disable this though. How about default initialize ```options.db_host_location``` to some well known string, like "default" or "hostname"? We can parse it in ```SanitizeOptions``` and replace it with the actual hostname. To disable the property completely, the user can just set ```options.db_host_location``` to an empty string.

File: db/db_impl/db_impl_open.cc [hunk 3]
```diff
@@ -181,6 +183,18 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
                    "file size check will be skipped during open.");
   }
 
+  if (result.db_host_location.empty()) {
+    result.db_host_location.resize(MAX_HOST_NAME_LEN, 0);
```
[the remark this answers is not in the document]
Comment (anand1976) [reply db/db_impl/db_impl_open.cc:187]: ```env->GetHostName()``` requires a pre-allocated ```char``` array.

File: db/db_impl/db_impl_open.cc [hunk 4]
```diff
@@ -181,6 +183,18 @@ DBOptions SanitizeOptions(const std::string& dbname, const DBOptions& src) {
...
+    if (!result.env
+             ->GetHostName(&result.db_host_location.at(0), MAX_HOST_NAME_LEN)
+             .ok()) {
+      ROCKS_LOG_INFO(result.info_log,
+                     "db_host_location property will not be set");
+    } else {
+      result.db_host_location.at(MAX_HOST_NAME_LEN - 1) = 0;
```
[the remark this answers is not in the document]
Comment (anand1976) [reply db/db_impl/db_impl_open.cc:194]: ```env->GetHostName()``` does not guarantee a null terminated string if the hostname exceeds the input array length.

File: HISTORY.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (siying) [reply HISTORY.md:11]: Can it be something like default value to be "___automatic___"?

[the remark this answers is not in the document]
Comment (mrambacher) [reply HISTORY.md:11]: How about this as a solution:
- We do not sanitize the DBOption but leave it "as is"
- If the "host_location" is empty, nothing is written to the SST file
- If the "host_location" is "localhost" or "default", the host name is written to the SST file (the Env is available in the options)

Also, it is not really a "location" but more of a "name" or "id".  

I still do not understand the use case for this location parameter and why I would want to use it.  What about the other table formats?

[the remark this answers is not in the document]
Comment (anand1976) [reply HISTORY.md:11]: Both sound like good ideas. Personally I like "__hostname__" or something like that as the "__" hints that it will be substituted with something, and "__hostname__" makes it clear what that something is. I think there's little risk of anyone taking it literally.

Regarding the option name, maybe "db_host_id" sounds better.

[the remark this answers is not in the document]
Comment (anand1976) [reply HISTORY.md:11]: > How about this as a solution:
> 
> * We do not sanitize the DBOption but leave it "as is"

That's a good point. That way the OPTIONS file can be read back in and validated without issues (which may have been your original concern).

File: options/options_helper.cc [hunk 5]
```diff
@@ -164,6 +164,7 @@ DBOptions BuildDBOptions(const ImmutableDBOptions& immutable_db_options,
       immutable_db_options.max_bgerror_resume_count;
   options.bgerror_resume_retry_interval =
       immutable_db_options.bgerror_resume_retry_interval;
+  //  options.db_host_location = immutable_db_options.db_host_location;
```
Comment (pdillinger) [inline options/options_helper.cc:167]: stale code?

File: table/format.cc [hunk 6]
```diff
@@ -403,4 +404,17 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
                                                    ioptions, allocator);
 }
 
+Status SanitizeDbHostIdProperty(Env* env, std::string& db_host_id) {
```
Comment (pdillinger) [inline table/format.cc:407]: Google style: use pointer for mutable parameters

File: table/format.cc [hunk 7]
```diff
@@ -403,4 +404,17 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
...
+  if (db_host_id == kDbHostId) {
+    db_host_id.clear();
+    db_host_id.resize(kMaxHostNameLen);
+    Status s = env->GetHostName(&db_host_id.at(0), kMaxHostNameLen);
+    if (!s.ok()) {
+      return s;
```
Comment (pdillinger) [inline table/format.cc:413]: Doesn't this leave the db_host_id non-empty?

File: table/format.cc [hunk 8]
```diff
@@ -403,4 +404,17 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
                                                    ioptions, allocator);
 }
 
+Status SanitizeDbHostIdProperty(Env* env, std::string& db_host_id) {
+  if (db_host_id == kDbHostId) {
+    db_host_id.clear();
+    db_host_id.resize(kMaxHostNameLen);
```
Comment (pdillinger) [inline table/format.cc:410]: I also find this extremely weird, to the point that you seem to have bugs yourself: using a fixed-length std::string to contain a \0 terminated C-style string. Performance should not be a concern to copy from a stack allocated fixed-size buffer to a std::string.

File: table/format.cc [hunk 9]
```diff
@@ -403,4 +404,17 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
...
+    Status s = env->GetHostName(&db_host_id.at(0), kMaxHostNameLen);
+    if (!s.ok()) {
+      return s;
+    }
+    db_host_id.at(kMaxHostNameLen - 1) = 0;
```
Comment (pdillinger) [inline table/format.cc:415]: Use '\0' for clarity for null termination character

File: table/format.cc [hunk 10]
```diff
@@ -403,4 +404,17 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
                                                    ioptions, allocator);
 }
 
+Status SanitizeDbHostIdProperty(Env* env, std::string& db_host_id) {
+  if (db_host_id == kDbHostId) {
+    db_host_id.clear();
+    db_host_id.resize(kMaxHostNameLen);
+    Status s = env->GetHostName(&db_host_id.at(0), kMaxHostNameLen);
```
Comment (pdillinger) [inline table/format.cc:411]: If GetHostName is really allowed to not include a terminating nul character, that should be documented in the public API. And best I can tell, the only other place we call this function does not account for that presumed behavior, I guess inherited from posix gethostname. Can't we do something a bit more sane, even without changing the Env function declaration?

File: include/rocksdb/options.h [hunk 11]
```diff
@@ -348,6 +348,8 @@ struct DbPath {
   DbPath(const std::string& p, uint64_t t) : path(p), target_size(t) {}
 };
 
+static const std::string kDbHostId = "__hostname__";
```
Comment (pdillinger) [inline include/rocksdb/options.h:351]: This name doesn't leave any room for alternatives. How about `kHostnameForDbHostId`?

File: table/format.h [hunk 12]
```diff
@@ -331,6 +331,11 @@ extern Status UncompressBlockContentsForCompressionType(
     BlockContents* contents, uint32_t compress_format_version,
     const ImmutableCFOptions& ioptions, MemoryAllocator* allocator = nullptr);
 
+static const size_t kMaxHostNameLen = 256;
+
+// Replace db_host_id contents with the real hostname if necessary
+extern Status SanitizeDbHostIdProperty(Env* env, std::string& db_host_id);
```
Comment (pdillinger) [inline table/format.h:337]: Maybe better to call this `ReifyDbHostIdProperty`

File: db/db_impl/db_impl_open.cc [hunk 13]
```diff
@@ -23,6 +23,8 @@
 #include "test_util/sync_point.h"
 #include "util/rate_limiter.h"
 
+#define MAX_HOST_NAME_LEN 256
```
[the remark this answers is not in the document]
Comment (pdillinger) [reply db/db_impl/db_impl_open.cc:26]: Appears stale / unused

File: HISTORY.md [hunk 14]
```diff
@@ -8,6 +8,7 @@
...
 * AdvancedColumnFamilyOptions::force_consistency_checks is now true by default, for more proactive DB corruption detection at virtually no cost (estimated two extra CPU cycles per million on a major production workload). Corruptions reported by these checks now mention "force_consistency_checks" in case a false positive corruption report is suspected and the option needs to be disabled (unlikely). Since existing column families have a saved setting for force_consistency_checks, only new column families will pick up the new default.
+* A new option, db_host_id, has been added to DBOptions. The option defaults to the hostname (FQDN). if set to something else,  the user specified string is written to SST files as a table property. If its empty, the property is not written.
```
Comment (pdillinger) [inline HISTORY.md:11]: Myself coming from a case of people not fully digesting release notes and their potential impacts...

I suggest starting with the fact that SST files now include hostnames, but an option is added to override/exclude that.

**Review by pdillinger — CHANGES_REQUESTED** [review]
Some things left to fix

File: table/format.cc (hunk 7 above)
Comment (anand1976) [reply table/format.cc:413]: Good point.

File: table/format.cc (hunk 8 above)
Comment (anand1976) [reply table/format.cc:410]: Yes, a temporary buffer is a better idea.

File: table/format.cc (hunk 10 above)
Comment (anand1976) [reply table/format.cc:411]: The current behavior provides an indication that the input buffer length is insufficient, so the caller can retry if desired. Changing the behavior would eliminate that flexibility. I would just document the API clearly, and maybe provide a wrapper inside RocksDB for simplicity.

File: table/plain/plain_table_builder.cc [hunk 15]
```diff
@@ -100,6 +100,10 @@ PlainTableBuilder::PlainTableBuilder(
   properties_.column_family_name = column_family_name;
   properties_.db_id = db_id;
   properties_.db_session_id = db_session_id;
+  properties_.db_host_id = ioptions.db_host_id;
+  if (!ReifyDbHostIdProperty(ioptions_.env, &properties_.db_host_id).ok()) {
+    ROCKS_LOG_WARN(ioptions_.info_log, "db_host_id property will not be set");
+  }
```
Comment (mrambacher) [inline table/plain/plain_table_builder.cc:106]: Warning seems pretty harsh for something the user chose to do...

File: table/sst_file_writer.cc [hunk 16]
```diff
@@ -7,6 +7,7 @@
 
 #include <vector>
 
+#include "db/db_impl/db_impl.h"
```
Comment (mrambacher) [inline table/sst_file_writer.cc:10]: Why is this necessary?

File: include/rocksdb/options.h [hunk 17]
```diff
@@ -1180,6 +1182,18 @@ struct DBOptions {
...
+  // will be written as a property in every SST file written by the DB (or
+  // by offline writers such as SstFileWriter and RepairDB). It can be useful
+  // for troubleshooting in memory corruption caused by a failing host when
+  // writing a file, by tracing back to the writing host. These corruptions
+  // may not be caught by the checksum since they happen before checksumming.
+  // If left as default, the table writer will substitute it with the actual
+  // hostname when writing the SST file.
+  //
+  // Default: hostname
+  std::string db_host_id = kHostnameForDbHostId;
```
[the remark this answers is not in the document]
Comment (anand1976) [reply include/rocksdb/options.h:1196]: It feels more intuitive to just set it to an empty string. Good point about the test. Let me add one.

File: options/options_settable_test.cc [hunk 18]
```diff
@@ -226,6 +226,7 @@ TEST_F(OptionsSettableTest, DBOptionsAllFieldsSettable) {
       {offsetof(struct DBOptions, wal_filter), sizeof(const WalFilter*)},
       {offsetof(struct DBOptions, file_checksum_gen_factory),
        sizeof(std::shared_ptr<FileChecksumGenFactory>)},
+      {offsetof(struct DBOptions, db_host_id), sizeof(std::string)},
```
[the remark this answers is not in the document]
Comment (anand1976) [reply options/options_settable_test.cc:229]: It is added to the test string. This line here is to skip filling the option address with garbage, which would cause a segfault.

File: table/sst_file_writer.cc (hunk 16 above)
Comment (anand1976) [reply table/sst_file_writer.cc:10]: Not needed anymore, since we don't need to sanitize options here anymore.

File: include/rocksdb/options.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (mrambacher) [reply include/rocksdb/options.h:1196]: I am fine with empty meaning "do not write one", but that should be part of the comment here.

File: table/format.cc [hunk 19]
```diff
@@ -403,4 +404,27 @@ Status UncompressBlockContents(const UncompressionInfo& uncompression_info,
                                                    ioptions, allocator);
 }
 
+Status GetHostName(Env* env, std::string* result) {
+  char hostname_buf[kMaxHostNameLen];
```
[the remark this answers is not in the document]
Comment (anand1976) [reply table/format.cc:408]: How would you access the storage though? You'll still need a temporary buffer. In any case, performance is not a concern.

File: include/rocksdb/env.h [hunk 20]
```diff
@@ -452,9 +452,15 @@ class Env {
...
 
-  // Get the current host name.
+  // Get the current host name as a null terminated string iff the string
+  // length is < len. The hostname should otherwise be truncated to len.
   virtual Status GetHostName(char* name, uint64_t len) = 0;
 
+  // Get the current hostname from the given env as a null terminated
+  // string in result. The result may be truncated if the hostname is too
+  // long
+  static Status GetHostName(Env* env, std::string* result);
```
Comment (pdillinger) [inline include/rocksdb/env.h:462]: It seems a little weird to me making this static. If making it non-static, virtual or not, you could use the name `GetHostNameString` to avoid compiler issues in combining overloading and overriding. Up to you.

File: table/sst_file_reader_test.cc [hunk 21]
```diff
@@ -90,6 +90,7 @@ class SstFileReaderTest : public testing::Test {
     if (check_global_seqno) {
       auto properties = reader.GetTableProperties();
       ASSERT_TRUE(properties);
+      ASSERT_FALSE(properties->db_host_id.empty());
```
Comment (pdillinger) [inline table/sst_file_reader_test.cc:93]: I'm not a fan of "copying the code into the unit test" as an oracle, but it might be worth comparing against GetHostName on the Env here. (I'm not too concerned about the case of someone changing their host name during this test case running.) Up to you

**Review by pdillinger — APPROVED** [review]
👍
